### PR TITLE
Fix ANR probability when user configures run conditions (fixes #1196)

### LINKED ISSUE #1196: ANR probability during run conditions UI stress test
There is an ANR problem already on my bugs-to-fix todo list. The ANR happens when you stress test the run conditions UI and change the checkboxes too quickly one after another. Since we've synchronized stuff to wipe out race crashes, the ANR is more suspect to happen. GPlay console proves this by a slightly higher ANR rate since app version 0.10.11.

Maybe related to issue: #1193 "Run conditions disregarded"

### Version Information
    App Version: 0.10.12, 0.10.11
    Syncthing Version: v0.14.49
    Android Version: Android 7.1.2

### PR DESCRIPTION
**Pull request by ghost**
Purpose
There is an ANR problem already on my bugs-to-fix todo list. The ANR happens when you stress test the run conditions UI and change the checkboxes too quickly one after another. Since we've synchronized stuff to wipe out race crashes, the ANR is more suspect to happen. GPlay console proves this by a slightly higher ANR rate since app version 0.10.11.

Related issue: #1196 
Maybe related to issue: #1193 "Run conditions disregarded"

Testing
Verified working on device lg-h815 running Android 7.1.2 at commit https://github.com/syncthing/syncthing-android/pull/1197/commits/28f2da26dabdf5d834b6796aa0dd9622b3b8197e  .

### CODE REVIEW

File: app/src/main/java/com/nutomic/syncthingandroid/activities/SettingsActivity.java [hunk 1]
```diff
@@ -385,6 +417,11 @@ public void onStop() {
                     mPendingConfig = false;
                 }
             }
+            if (mPendingRunConditions) {
+                if (mSyncthingService != null) {
+                    mSyncthingService.reEvaluateRunConditions();
```
[the remark this answers is not in the document]
Comment (Catfriend1) [reply app/src/main/java/com/nutomic/syncthingandroid/activities/SettingsActivity.java]: Ok, agreed. https://github.com/syncthing/syncthing-android/pull/1197/commits/0328463e794cfb9713634c0699186db6d605db8a

**Comment by AudriusButkevicius** [comment]
I don't think this causes an ANR.

**Comment by Catfriend1** [comment]
@AudriusButkevicius Want proof? I don't want to sound impolite but try it. We have many reports on GPlay Console that users experience ANR and in daily use or testing, I don't have them except I stress test the run conditions prefs UI.

**Comment by AudriusButkevicius** [comment]
As I said in the other thread, I don't see why race conditions on some checkboxes or class values would cause the application to hang.

If you are able to reproduce the hang, then run with the debugger attached, see what the UI thread is doing, explaining exactly what condition causes this.

For now it just seems that you did something, you think/hope that it went away, but can't explain what actually causes it.

**Comment by Catfriend1** [comment]
I did this. ( If you are able to reproduce the hang, then run with the debugger attached, see what the UI thread is doing, explaining exactly what condition causes this. )

Explained it on the issue ( For now it just seems that you did something, you think/hope that it went away, but can't explain what actually causes it.  )

**Comment by Catfriend1** [comment]
Doing the analysis on my phone, I get a similar log output. It's always stuck during waiting multiple seconds for the safe shutdown with a high chance, and stuck on executing libsyncthing.so with a low chance (as the exit routine takes more time with the sleep(x) seconds). This is a problem even before 0.10.x and 0.9.x because the changed prefs trigger multiple(!) immediate evaluations of the run conditions instead of triggering once(!) after leaving the prefs screen.
You'll see each log a little different as you don't know when android's os-side ANR timeout is reached.

Log from Gplay Console:

> "main" prio=5 tid=1 Waiting
  | group="main" sCount=1 dsCount=0 flags=1 obj=0x727e6970 self=0xa45f3000
  | sysTid=16091 nice=0 cgrp=default sched=0/0 handle=0xa7e904a4
  | state=S schedstat=( 2822983226 658785428 4322 ) utm=260 stm=21 core=0 HZ=100
  | stack=0xbe456000-0xbe458000 stackSize=8MB
  | held mutexes=
  at java.lang.Object.wait (Native method)
- waiting on <0x00f3c691> (a java.lang.Object)
  at java.lang.Object.wait (Object.java:422)
  at java.lang.Thread.join (Thread.java:1262)
- locked <0x00f3c691> (a java.lang.Object)
  at java.lang.Thread.join (Thread.java:1338)
  at com.nutomic.syncthingandroid.service.SyncthingService.shutdown (SyncthingService.java:506)
  at com.nutomic.syncthingandroid.service.SyncthingService.onDestroy (SyncthingService.java:459)
- locked <0x0371e9f6> (a java.lang.Object)
  at android.app.ActivityThread.handleStopService (ActivityThread.java:3505)
  at android.app.ActivityThread.-wrap26 (ActivityThread.java)
  at android.app.ActivityThread$H.handleMessage (ActivityThread.java:1698)
  at android.os.Handler.dispatchMessage (Handler.java:106)
  at android.os.Looper.loop (Looper.java:164)
  at android.app.ActivityThread.main (ActivityThread.java:6499)
  at java.lang.reflect.Method.invoke (Native method)
  at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run (RuntimeInit.java:440)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:807)

**Comment by AudriusButkevicius** [comment]
Why would people repeatedly jam the options?

I repeatedly jam the options on my phone (always run checked, jamming run only on charge while not charging), and nothing happens.

**Comment by Catfriend1** [comment]
@AudriusButkevicius It happens when you use a Samsung Galaxy S2 (device i9100) phone which is pretty slow in loading and terminating the arm_x86 libsyncthing.so or if you use a modern phone under heavy load by another app. Then the allowed time frame is consumed up. People cause the permanent start/stops of libsyncthing by fastly going into the run conditions screen and ticking from top to down:
- always run in background
- run only when charging
- run only on wifi
Correct preconditions in reality given this causes on>off, if the user checks the wrong option we expect him to correct that "fastly". Anyways, strategically, we apply syncthing options (like address and protocol) on leaving the settings screen, the run conditions should also apply then and not immediately when ticking the box.

**Comment by AudriusButkevicius** [comment]
So I don't disagree with the change, the change is local and correct. I just find it hard to believe that this fixes. Perhaps I am wrong, perhaps I am write, but regardless of that the change is ok.

**Review by AudriusButkevicius — APPROVED** [review]

**Pull request merged**